[BUG] NPM v7 uses SSH instead of an explicit HTTPS for GitHub repos

<!--
Note: Please search to see if an issue already exists for your problem: https://github.com/npm/cli/issues
-->

### Current Behavior:
<!-- ex. a clear & concise description of what you're experiencing. -->

When I use a git repository via an HTTP link NPM "takes liberties" with it, which breaks my build:

```txt
$ npm init -y
Wrote to /Users/eugene.lazutkin/Work/temp/package.json:

{
  "name": "temp",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}


$ npm i --save https://github.com/uhop/stream-chain.git

added 1 package, and audited 2 packages in 3s

found 0 vulnerabilities
```

It produces `package-lock.json`:

```json
{
  "name": "temp",
  "version": "1.0.0",
  "lockfileVersion": 2,
  "requires": true,
  "packages": {
    "": {
      "version": "1.0.0",
      "license": "ISC",
      "dependencies": {
        "stream-chain": "github:uhop/stream-chain"
      }
    },
    "node_modules/stream-chain": {
      "version": "2.2.4",
      "resolved": "git+ssh://git@github.com/uhop/stream-chain.git#459f5a1708c138b6e0abaae4cf103c3488e1e78e",
      "license": "BSD-3-Clause"
    }
  },
  "dependencies": {
    "stream-chain": {
      "version": "git+ssh://git@github.com/uhop/stream-chain.git#459f5a1708c138b6e0abaae4cf103c3488e1e78e",
      "from": "stream-chain@github:uhop/stream-chain"
    }
  }
}
```

Note that `https://github.com/uhop/stream-chain.git` was replaced with `github:uhop/stream-chain`, which is probably OK in this case. But other two links (?) are rewritten from `https://github.com/uhop/stream-chain.git` to `git+ssh://git@github.com/uhop/stream-chain.git`, which is clearly bad.

The problem is that a build bot we use in similar situations can access private git repositories using HTTP, but not SSH for security reasons. It fails on an authentication. Rewriting `https://github.com/uhop/stream-chain.git` to `git+ssh://git@github.com/uhop/stream-chain.git` is not acceptable for that reasons.

The fix is relatively minor yet unpleasant: we have to replace `npm ci` with `npm i`, which takes more time and introduced instabilities with other dependencies.

### Expected Behavior:
<!-- ex. a clear & concise description of what you expected to happen. -->

When running `npm ci` it should use the original URL with the HTTP authentication instead of SSH.

### Steps To Reproduce:
<!--
ex. steps to reproduce the behavior:
1. In this environment...
2. With this config...
3. Run '...'
4. See error...
-->

See the description and do the same steps using git repositories (github only?) as dependencies.

### Environment:
<!--
ex.
- OS: Ubuntu 20.04
- Node: 13.14.0
- npm: 6.4.12
-->

OS: Mac
Node: 15.7.0
NPM: 7.4.3

This is a breaking change in npm v7 and prevents anyone using even **public** packages without ssh keys.

We have made a fork of an npm package and host it as a public repo.  We refer to it using a `github:` link.  In v6 this github link was populated to package-lock.json as well, which worked fine.  In v7 it is converted into `git+ssh:` which fails in our CI/CD pipeline as it doesn't have any ssh keys configured.

We don't want to add ssh keys to our CI pipeline just to fetch **public** packages, so I don't see any way we can upgrade.  I'd be glad to hear of any workaround.

Specific example:

In package.json referring to https://github.com/Vincit/winston-azure-transport:
```
    "winston-azure-transport": "github:Vincit/winston-azure-transport#ed07d0685ca601638ab8ebcf660128848dd30215"
```
Diff of package-lock.json between npm v6 and v7:
```diff
     "winston-azure-transport": {
-      "version": "github:Vincit/winston-azure-transport#ed07d0685ca601638ab8ebcf660128848dd30215",
-      "from": "github:Vincit/winston-azure-transport#ed07d0685ca601638ab8ebcf660128848dd30215",```
+      "version": "git+ssh://git@github.com/Vincit/winston-azure-transport.git#ed07d0685ca601638ab8ebcf660128848dd30215",
+      "from": "winston-azure-transport@github:Vincit/winston-azure-transport#ed07d0685ca601638ab8ebcf660128848dd30215",
```
Our CI now fails with:
```
npm ERR! Error while executing:
npm ERR! /usr/bin/git ls-remote -h -t ssh://git@github.com/Vincit/winston-azure-transport.git
npm ERR! 
npm ERR! git@github.com: Permission denied (publickey).
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.
npm ERR! 
npm ERR! exited with error code: 128
```

--- a/bin/check-latest-npm.js
+++ b/bin/check-latest-npm.js
@@ -12,7 +12,9 @@ const semver = require( 'semver' );
  * Internal dependencies
  */
 const {
-	engines: { npm: npmSemver },
+	engines: { npm: npmRange },
+	// Ignore reason: `package.json` exists outside `bin` `rootDir`.
+	// @ts-ignore
 } = require( '../package.json' );
 
 /**
@@ -60,7 +62,7 @@ async function getLatestNPMVersion() {
 
 				const versions = Object.values( data[ 'dist-tags' ] );
 
-				resolve( semver.maxSatisfying( versions, npmSemver ) );
+				resolve( semver.maxSatisfying( versions, npmRange ) );
 			}
 		).on( 'error', ( error ) => {
 			if (

--- a/bin/check-latest-npm.js
+++ b/bin/check-latest-npm.js
@@ -6,12 +6,20 @@
 const { green, red, yellow } = require( 'chalk' );
 const { get } = require( 'https' );
 const { spawn } = require( 'child_process' );
+const semver = require( 'semver' );
+
+/**
+ * Internal dependencies
+ */
+const {
+	engines: { npm: npmSemver },
+} = require( '../package.json' );
 
 /**
  * Returns a promise resolving with the version number of the latest available
- * version of NPM.
+ * version of npm.
  *
- * @return {Promise<string>} Promise resolving with latest NPM version.
+ * @return {Promise<string>} Promise resolving with latest npm version.
  */
 async function getLatestNPMVersion() {
 	return new Promise( ( resolve, reject ) => {
@@ -30,7 +38,7 @@ async function getLatestNPMVersion() {
 			async ( response ) => {
 				if ( response.statusCode !== 200 ) {
 					return reject(
-						new Error( 'Package data for NPM not found' )
+						new Error( 'Package data for npm not found' )
 					);
 				}
 
@@ -45,23 +53,25 @@ async function getLatestNPMVersion() {
 				} catch {
 					return reject(
 						new Error(
-							'Package data for NPM returned invalid response body'
+							'Package data for npm returned invalid response body'
 						)
 					);
 				}
 
-				resolve( data[ 'dist-tags' ].latest );
+				const versions = Object.values( data[ 'dist-tags' ] );
+
+				resolve( semver.maxSatisfying( versions, npmSemver ) );
 			}
 		).on( 'error', ( error ) => {
 			if (
 				/** @type {NodeJS.ErrnoException} */ ( error ).code ===
 				'ENOTFOUND'
 			) {
-				error = new Error( `Could not contact the NPM registry to determine latest version.
+				error = new Error( `Could not contact the npm registry to determine latest version.
 
 This could be due to an intermittent outage of the service, or because you are not connected to the internet.
 
-Because it is important that \`package-lock.json\` files only be committed while running the latest version of NPM, this commit has been blocked.
+Because it is important that \`package-lock.json\` files only be committed while running the latest version of npm, this commit has been blocked.
 
 If you are certain of your changes and desire to commit anyways, you should either connect to the internet or bypass commit verification using ${ yellow(
 					'git commit --no-verify'
@@ -75,9 +85,9 @@ If you are certain of your changes and desire to commit anyways, you should eith
 
 /**
  * Returns a promise resolving with the version number of the local installed
- * version of NPM.
+ * version of npm.
  *
- * @return {Promise<string>} Promise resolving with local installed NPM version.
+ * @return {Promise<string>} Promise resolving with local installed npm version.
  */
 async function getLocalNPMVersion() {
 	return new Promise( async ( resolve ) => {
@@ -99,15 +109,15 @@ Promise.all( [ getLatestNPMVersion(), getLocalNPMVersion() ] )
 	.then( ( [ latest, local ] ) => {
 		if ( latest !== local ) {
 			throw new Error(
-				`The local NPM version does not match the expected latest version. Expected ${ green(
+				`The local npm version does not match the expected latest version. Expected ${ green(
 					latest
 				) }, found ${ red( local ) }.
 
-It is required that you have the latest version of NPM installed in order to commit a change to the package-lock.json file.
+It is required that you have the expected latest version of npm installed in order to commit a change to the package-lock.json file.
 
 Run ${ yellow(
-					'npm install --global npm@latest'
-				) } to install the latest version of NPM. Before retrying your commit, run ${ yellow(
+					`npm install --global npm@${ latest }`
+				) } to install the expected latest version of npm. Before retrying your commit, run ${ yellow(
 					'npm install'
 				) } once more to ensure the package-lock.json contents are correct. If there are any changes to the file, they should be included in your commit.`
 			);
@@ -115,7 +125,7 @@ Run ${ yellow(
 	} )
 	.catch( ( error ) => {
 		console.error(
-			'Latest NPM check failed!\n\n' + error.toString() + '\n'
+			'Latest npm check failed!\n\n' + error.toString() + '\n'
 		);
 		process.exitCode = 1;
 	} );

--- a/package.json
+++ b/package.json
@@ -16,7 +16,7 @@
 	},
 	"engines": {
 		"node": ">=10.0.0",
-		"npm": ">=6.9.0"
+		"npm": ">=6.9.0 <7"
 	},
 	"config": {
 		"GUTENBERG_PHASE": 2,

